doc/developer/design/20230512_mir_cost_model.md

```diff
@@ -0,0 +1,149 @@
+- Collect all relevant statistics in storage and make them available to the cost model.
+
+# Explanation
+[explanation]: #explanation
+
+## Proposal
+
+A cost model is an abstract interpretation of the query: rather than running the query on relations and getting new relation out, we run the query on a symbolic description of its input relations and get a symbolic description of the result out.
+
+The general approach can be the same as `mz_transform::Typecheck`: recursively walk the `MirRelationExpr` with some contextual information, returning a symbolic term characterizing the query.
+
+### Context information
+
+Traditional databases track cardinality data and statistics about various columns. We will be interested in other statistics, too. Some possible examples:
+
+- How frequently does data arrive?
+- Is the given input physically monotonic or append-only? (I.e., there are no retractions.)
+- At what rate do we expect new data to arrive?
+- Some kind of locality property?
+
+### Symbolic costs
+
+A symbolic cost is a polynomial over the features of the input relations. For example, a batch query that loops over relation `R` twice and then computes its cross product with relation `Q` will have cost `2|R| + |R|*|Q|`. A traditional database typically computes cost just in terms of overall batch query time, but our costs are multi-dimensional. In rough feasibility order:
+
+| Dimension          | Input features                                               | Outputs                         |
+| ------------------ | ------------------------------------------------------------ | ------------------------------- |
+| Rehydration time   | Source cardinality                                           | Time estimate                   |
+| Latency per update | Source cardinality, expected per-update delta on each source | Time estimate (per source)      |
+| Storage            | Source cardinality, existing indices, cardinality of groups  | Arrangements and size estimates |
+| Scalability        | Distribution of keys (e.g., cardinality of distinct keys)    | Data parallelism estimate       |
+
+The *rehydration time* dimension is effectively a classical batch database cost model: given cardinalities on the input relations, predict how much work a given plan has to do.
+
+The *latency per update* dimension is unique to our streaming setting. Different query plans will lead to different costs when different sources produce new data. If a calculus metaphor is helpful, you can think of the cost model here producing a partial derivative with respect to each input; alternatively, the cost model’s aggregate *latency per update* cost is the [Jacobian](https://en.wikipedia.org/wiki/Jacobian_matrix_and_determinant) of that query.
+
+The *storage* dimension is meant to identify particularly expensive new arrangements: for example, a count query grouping by US state code (just 50, no big deal) vs. grouping by user ID (as big as your user table).
```

Comment:
There are now also different alternatives for some operators that can impact arrangement storage costs. For example, we now do not use the `Plan::DistinctNegated` physical operator, even though it should in some situations be better than `Plan::Distinct` (see [Slack thread](https://materializeinc.slack.com/archives/C02PPB50ZHS/p1676288759455119)).

Like in the monotonicity discussion above, there is again here some need to consider the division of responsibility between MIR and LIR. It seems to be the case that operator selection should happen at LIR, but operator selection affects the actual storage costs that are to be estimated at MIR.

Reply:
Right---this came up in discussion of join planning, too.

Without a strict contract at the MIR/LIR level, we need to a bit of fortune telling to happen at the MIR level to compute a cost. Such predictions of the future are error prone, because they duplicate logic: we have the actual logic that translates to LIR and we have the cost model which makes predictions at the MIR level, and these two bits of code can drift apart if we're not careful. I don't know a good way to avoid the duplicate work, though.

Reply:
I think this makes sense, especially because we are not proposing to revisit optimizer architecture as part of this work. There are DBMS where the interaction between physical optimization vs. transformation modules is [different](https://www.researchgate.net/profile/Rafi-Ahmed-2/publication/221311318_Cost-Based_Query_Transformation_in_Oracle/links/572bbc5e08aef7c7e2c6b829/Cost-Based-Query-Transformation-in-Oracle.pdf), in that there is more fluidity allowing for more information exchange. Since we have a single lowering pass from MIR to LIR, we do not have the same flexibility (but also none of the corresponding challenges that come with it).

Reply:
But in the meantime, the join ordering will look at only cardinalities, which should be ok to compute on MIR, right?